cpp/bench/sg/rf_classifier.cu

```diff
@@ -0,0 +1,139 @@
+/*
+ * Copyright (c) 2019, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cuML.hpp>
+#include <randomforest/randomforest.hpp>
+#include <utility>
+#include "benchmark.cuh"
+
+namespace ML {
+namespace Bench {
+namespace rf {
+
+struct Params {
+  DatasetParams data;
+  BlobsParams blobs;
+  RF_params rf;
+};
+
+template <typename D>
+struct RFClassifierModel {};
+
+template <>
+struct RFClassifierModel<float> {
+  ML::RandomForestClassifierF model;
+};
+
+template <>
+struct RFClassifierModel<double> {
+  ML::RandomForestClassifierD model;
+};
+
+template <typename D>
+class RFClassifier : public BlobsFixture<D> {
+ public:
+  RFClassifier(const std::string& name, const Params& p)
+    : BlobsFixture<D>(p.data, p.blobs), rfParams(p.rf) {
+    this->SetName(name.c_str());
+  }
+
+ protected:
+  void runBenchmark(::benchmark::State& state) override {
+    if (this->params.rowMajor) {
+      state.SkipWithError("RFClassifier only supports col-major inputs");
+    }
+    auto& handle = *this->handle;
+    auto stream = handle.getStream();
+    auto* mPtr = &model.model;
+    for (auto _ : state) {
+      CudaEventTimer timer(handle, state, true, stream);
+      mPtr->trees = nullptr;
+      fit(handle, mPtr, this->data.X, this->params.nrows, this->params.ncols,
+          labels, this->params.nclasses, rfParams);
+      CUDA_CHECK(cudaStreamSynchronize(stream));
+      delete[] mPtr->trees;
+    }
+  }
+
+  void allocateBuffers(const ::benchmark::State& state) override {
+    auto allocator = this->handle->getDeviceAllocator();
+    auto stream = this->handle->getStream();
+    labels =
+      (int*)allocator->allocate(this->params.nrows * sizeof(int), stream);
+  }
+
+  void deallocateBuffers(const ::benchmark::State& state) override {
+    auto allocator = this->handle->getDeviceAllocator();
+    auto stream = this->handle->getStream();
+    allocator->deallocate(labels, this->params.nrows * sizeof(int), stream);
+  }
+
+ private:
+  int* labels;
+  RFClassifierModel<D> model;
+  RF_params rfParams;
+};
+
+template <typename D>
+std::vector<Params> getInputs() {
```

Comment:
What would be really amazing would be if we could read this from a json file. That way we could provide a series of json files in `cpp/bench/conf`and accept them in the command-line. That way we could just build a fixture for each different supported dataset and execute them at runtime depending on the contents of the configuration file. Would make things super easy. 

Now that I think about it, we could even give this same support to the Python API, and be able to use the same json / config files across both Python & C++. This would certainly make bug finding and stress testing much easier across the board. What do you think?

Reply:
this is a cool idea! I have filed issue #1156 for this. IMO, this has a bigger impact to our whole benchmarking strategy and thus we should try to plan it for our 0.11 release.